packages/backend-for-frontend/src/services/toolService.ts

```diff
@@ -0,0 +1,469 @@
+        purposeId,
+      },
+    };
+  }
+
+  const purpose = await purposeProcessClient.getPurpose({
+    params: { id: purposeId },
+    headers: ctx.headers,
+  });
+
+  const agreement = await retrieveAgreement(
+    agreementProcessClient,
+    purpose.consumerId,
+    purpose.eserviceId,
+    ctx
+  );
+
+  const descriptor = await retrieveDescriptor(
+    catalogProcessClient,
+    agreement.eserviceId,
+    agreement.descriptorId,
+    ctx
+  );
+
+  return {
+    errors: undefined,
+    data: {
+      clientKind: authorizationApi.ClientKind.enum.CONSUMER,
+      clientId: unsafeBrandId<ClientId>(jwt.payload.iss),
+      kid: jwt.header.kid,
+      algorithm: "RS256",
+      publicKey: encodedPem,
+      purposeId,
+      consumerId: unsafeBrandId<TenantId>(keyWithClient.client.consumerId),
+      agreementId: unsafeBrandId<AgreementId>(agreement.id),
+      eServiceId: unsafeBrandId<EServiceId>(agreement.eserviceId),
+      agreementState: agreementStateToItemState(agreement.state),
+      purposeState: retrievePurposeItemState(purpose),
+      descriptorState: descriptorStateToItemState(descriptor.state),
+    },
+  };
+}
+
+async function retrieveAgreement(
+  agreementClient: AgreementProcessClient,
+  consumerId: string,
+  eserviceId: string,
+  ctx: WithLogger<BffAppContext>
+): Promise<agreementApi.Agreement> {
+  const agreements = await getAllAgreements(agreementClient, ctx.headers, {
+    consumersIds: [consumerId],
+    eservicesIds: [eserviceId],
+    states: [
+      agreementApi.AgreementState.Values.ACTIVE,
+      agreementApi.AgreementState.Values.SUSPENDED,
+    ],
+  });
+
+  assertOnlyOneAgreementForEserviceAndConsumerExists(
+    agreements,
+    eserviceId,
+    consumerId
+  );
+  return agreements[0];
+}
+
+async function retrieveDescriptor(
+  catalogClient: CatalogProcessClient,
+  eserviceId: string,
+  descriptorId: string,
+  ctx: WithLogger<BffAppContext>
+): Promise<catalogApi.EServiceDescriptor> {
+  const eservice = await catalogClient.getEServiceById({
+    params: { eServiceId: eserviceId },
+    headers: ctx.headers,
+  });
+
+  const descriptor = eservice.descriptors.find((d) => d.id === descriptorId);
+  if (!descriptor) {
+    throw eserviceDescriptorNotFound(eservice.id, descriptorId);
+  }
+
+  return descriptor;
+}
+
+function retrievePurposeItemState(purpose: purposeApi.Purpose): ItemState {
+  const activePurposeVersion = [...purpose.versions]
+    .sort(
+      (a, b) =>
+        new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
+    )
+    .find(
+      (v) =>
+        v.state === purposeApi.PurposeVersionState.Enum.ACTIVE ||
+        v.state === purposeApi.PurposeVersionState.Enum.SUSPENDED
+    );
+
+  if (!activePurposeVersion) {
+    throw missingActivePurposeVersion(purpose.id);
+  }
+
+  return purposeVersionStateToItemState(activePurposeVersion.state);
+}
+
+async function retrieveTokenValidationEService(
```

Comment:
this function and `retrieveKey` make the same call to retrieve the same values, can we re-use them?

Note: the agreement is retrieved using 2 different approaches